### ISSUE TITLE
Minor text adjustments after re-structuring API doc.  (was: New proposed structure for API doc)

### ISSUE THREAD

**Issue by mirjak** [ietf-tapswg/api-drafts#458]
I think we should reorganise the API doc a little bit. I proposed to 1) have the "Managing Connection" section earlier and then 2) have a new top level section on "Data Transfer" (inline with the arch doc) and have first one subsection that contains everything around messages and then having "Sending Data" and "Receiving Data" as two more subsections.

I think the only text changed ended is in the "Sending Data" section where currently messages are introduced as this structure would introduce messages earlier and some new text for the start of the new "Data Transfer" section of course.

The structure would then look like this:

7. Managing Connections 
     7.1.  Generic Connection Properties 
       7.1.1.  Retransmission Threshold Before Excessive Retransmission Notification 
       7.1.2.  Required Minimum Corruption Protection Coverage for Receiving 
       7.1.3.  Priority (Connection)
       7.1.4.  Timeout for Aborting Connection 
       7.1.5.  Connection Group Transmission Scheduler  
       7.1.6.  Maximum Message Size Concurrent with Connection Establishment 
       7.1.7.  Maximum Message Size Before Fragmentation or Segmentation 
       7.1.8.  Maximum Message Size on Send 
       7.1.9.  Maximum Message Size on Receive
       7.1.10. Capacity Profile 
       7.1.11. Bounds on Send or Receive Rate 
       7.1.12. TCP-specific Property: User Timeout 
     7.2.  Soft Errors
     7.3.  Excessive retransmissions
8.  Data Transfer
 8.1 Messages and Framers
   8.1.2.  Message Contexts  
   8.1.2. Message Framers 
     8.1.2.1.  Adding Message Framers to Connections 
     8.1.2.2.  Framing Meta-Data  
   8.1.3  Message Properties 
       8.1.3.1.  Lifetime  
       8.1.3.2.  Priority  
       8.1.3.3.  Ordered 
       8.1.3..4.  Idempotent 
       8.1.3.5.  Final 
       8.1.3.6.  Corruption Protection Length 
       8.1.3.7.  Reliable Data Transfer (Message) 
       8.1.3.8.  Message Capacity Profile Override 
       8.1.3.9.  Singular Transmission
 8.2 Sending Data 
     8.2.1.  Basic Sending
     8.2.2  Sending Replies 
     8.2.3.  Send Events 
       8.2.3.1.  Sent  
       8.2.3.2.  Expired 
       8.2.3.3.  SendError 
     8.2.4.  Partial Sends
     8.2.5.  Batching Sends  
     8.2.6.  Send on Active Open: InitiateWithSend 
 8.3  Receiving Data  
     8.3.1.  Enqueuing Receives  
     8.3.2.  Receive Events 
       8.3.2.1.  Received 
       8.3.2.2.  ReceivedPartial
       8.3.2.3.  ReceiveError  
     8.3.3.  Receive Message Properties 
       8.3.3.1.  ECN 
       8.3.3.2.  Early Data  
       8.3.3.3.  Receiving Final Messages

**Comment by theri** [comment]
I like this idea and I agree that Connection Properties should be introduced earlier.
Also, this structure introduces messageContexts before it uses them all the time.

One thing I'm wondering about, not actually related to the reordering, but while we're at it: 
Why is "TCP-specific Property: User Timeout" under "Generic Connection Properties" even though it's not a Generic but a Protocol Specific Property? Shouldn't this heading be on the same subsection level as "Generic Connection Property"?

**Comment by mirjak** [comment]
I have a separate PR that already changes that. I issued this proposed new structure before I filled the PR, so I didn't consider that. So yes!

**Comment by tfpauly** [comment]
I'd suggest keeping data transfer first, before managing

**Comment by mirjak** [comment]
This should be done after the next submitted version (when all PRs are cleared) and then get immediately submitted as another separate version.

**Comment by mwelzl** [comment]
I have now done this with commits to the master, to avoid blocking other updates with a long-standing PR. Please take a look - I think the structure is now correct, and the basic text is in place (as you say above, the intro to sending, and the intro to the new "Data Transfer" section). A few more text updates may be in order - e.g., in the new section 8.1.3 (Message Properties), this text now seems weird:

"The simpler form of Send, which does not take any messageContext,"
and should perhaps be written as: "If no Message Context is provided when sending, "  ... or something like that.

I think that sections 7 and 8 now just need to be read thoroughly by someone, and the text updated accordingly to avoid weird things like this due to the new structure.

This is still assigned to me - so I can do this, with a normal PR, next. Anyway, this issue is no longer in the way of other updates now.

**Comment by mirjak** [comment]
I thought the plan was to actually submit another version to the datertracker without the restructuring and then submit the restricted one right after, so it'd be easier to track changes there as well...?

**Comment by csperkins** [comment]
Right - I'd really have liked to see drafts submitted to the data tracker immediately before and after this refactor. It's very difficult to follow the changes on GitHub.

**Comment by mwelzl** [comment]
The previously submitted version is almost "right before" - there are only very few things that happened since then. If you want to submit the "immediately after" version now, please be my guest - I still have trouble with the tool chain for submission. Either way, what do you expect from this? See text shifts? It's not very interesting - it'll show you that things had the previous structure before, and the one above now. You won't see anything fascinating. If you're interested in what I've done beyond cut and paste, this shows it better:
https://github.com/ietf-tapswg/api-drafts/commit/f0f4a5d84c2b0d8d659974d60f6dc656bde6d6f4
(because it shows *just* that).

**Comment by csperkins** [comment]
The restructure maybe isn't very interesting, but it makes it obscures the other changes. Lots of people looks at diffs in the datatracker. That's much easier if a restructuring is done with a new draft version that only does the restructuring, and doesn't include technical changes.

And yes, I'd submit an update, but can't because the draft submissions are closed. Which is another reason why I thought we were waiting to do this restructuring.

**Comment by mwelzl** [comment]
Aha. Well sorry, I thought it doesn't matter when the pre-update changes are so small; I understand now, you say "lots of people", so the concern is probably about folks who might complain for having a mix of tech changes and the structure update. As for the timing, it seems I misunderstood this too - Mirja had said that she wants something to be submitted immediately after the submission system re-opens. I thought the idea was for me to do the structure update now such that we can submit the post-restructuring version immediately upon re-opening (which we still can, we just shouldn't merge any PRs until then).

Either way, I think that can be solved relatively easily by someone git-savvy. If the plan is to have a clean version submitted before re-structuring, then re-structure, then submit again, all we need to do is:
1) roll the master back to a commit just before the three that I did to re-structure
2) submit when the system is open,
3) re-apply the three commits,
4) submit again.

I can roll back to a commit before the one I applied, and keep a copy of my latest version so I can re-apply them easily after someone submitted and tells me "APPLY YOUR CHANGES NOW". Should I?

**Comment by mirjak** [comment]
Don't "roll back" the commit. I believe you can just check out an old version, build and tag that one for submission (on Sunday or Monday).

**Comment by csperkins** [comment]
What @mirjak said. We can reconstruct this from the Git history without rolling anything back, provided we know the git revision numbers.

**Comment by csperkins** [comment]
@mwelzl I always find it easiest to do a fresh git checkout of a particular version, rather than try to go back.

**Comment by tfpauly** [comment]
@mwelzl Going from a detached head is fine. Just check out that commit, build, and submit. No big deal. I can do that too.

**Comment by mwelzl** [comment]
Regarding the last comments: thanks; that's all fine, but we now have these two files - "draft-ietf-taps-interface_before_restructuring.md" and "draft-ietf-taps-interface_after_restructuring.md", we can just submit those and stop debating how to best do this   :)    the current master is up-to-date (after restructuring), so we can just continue with it.

**Comment by gorryfair** [comment]
YES Let's post the interim rev of the ID, and work on the next :-)

**Comment by mwelzl** [comment]
Yay, I did it!  Submitting wasn't easy, I had a string of error dependencies... in the end, fixing a brew permission error resolved things for me and I could compile and submit. I have now deleted the two temporary files, "draft-ietf-taps-interface_before_restructuring.md" and "draft-ietf-taps-interface_after_restructuring.md", from the repository.

**Comment by csperkins** [comment]
Thanks!

(And I know your pain – the markdown drafts are not easy to build…)

**Comment by mwelzl** [comment]
I gave the shifted sections a close read now, and found no more oddities. Now I think that even the text I quoted above ("The simpler form of Send, which does not take any messageContext,") actually fits where it is, because it elaborates on the example that's given above - basically saying that this is not the only way it could be done, the messageContext parameter could also be skipped.

So, I think the time has come to close this issue.
